services/statestorage/merkle/merkle_forest_test.go

```diff
@@ -29,137 +27,68 @@ import (
-	newRoot, _ := f.forest.Update(baseHash, diffs)
-	f.top++
-	f.roots[f.top] = newRoot
-	return newRoot
-}
-
-func updateStringEntries(f *driver, keyValues ...string) primitives.BlockHeight {
+func updateStringEntries(f *Forest, baseHash primitives.MerkleSha256, keyValues ...string) primitives.MerkleSha256 {
 	if len(keyValues)%2 != 0 {
 		panic("expected key value pairs")
 	}
-	var root primitives.MerkleSha256
+	currentRoot := baseHash
 	for i := 0; i < len(keyValues); i = i + 2 {
-		root = f.forest.updateSingleEntry(f.roots[f.top], keyValues[i], hash.CalcSha256([]byte(keyValues[i+1])))
-		f.top++
-		f.roots[f.top] = root
+		currentRoot = f.updateSingleEntry(currentRoot, keyValues[i], hash.CalcSha256([]byte(keyValues[i+1])))
 	}
-	return f.top
+	return currentRoot
 }
 
-func verifyProof(t *testing.T, f *driver, rootId primitives.BlockHeight, proof Proof, contract string, key string, value string, exists bool) {
-	rootHash, _ := f.getRootHash(rootId)
-	verified, err := f.forest.Verify(rootHash, proof, contract, key, value)
+func verifyProof(t *testing.T, f *Forest, root primitives.MerkleSha256, proof Proof, contract string, key string, value string, exists bool) {
+	verified, err := f.Verify(root, proof, contract, key, value)
 	require.NoError(t, err, "proof verification failed")
 	require.Equal(t, exists, verified, "proof verification returned unexpected result")
 }
 
-func getProofRequireHeight(t *testing.T, f *driver, rootId primitives.BlockHeight, contract string, key string, expectedHeight int) Proof {
-	root, err := f.getRootHash(rootId)
-	require.NoError(t, err, "failed getting root with error: %s", err)
-	proof, err := f.forest.GetProof(root, contract, key)
+func getProofRequireHeight(t *testing.T, f *Forest, root primitives.MerkleSha256, contract string, key string, expectedHeight int) Proof {
+	proof, err := f.GetProof(root, contract, key)
 	require.NoError(t, err, "failed with error: %s", err)
 	require.Len(t, proof, expectedHeight, "unexpected proof length")
 	return proof
 }
 
-func TestGetTopRootHash(t *testing.T) {
-	f := newForestDriver()
-
-	rootId := updateStringEntries(f, "first", "val")
-	topRoot, err1 := f.getTopRootHash()
-	updatedRoot, err2 := f.getRootHash(rootId)
-
-	require.NoError(t, err1, "GetTopHash failed with error")
-	require.NoError(t, err2, "getRootHash failed with error")
-	require.Equal(t, updatedRoot, topRoot, "getTopRootHash did not match getRootHash with rootId %v", rootId)
-}
-
-func TestGetPastRootHash(t *testing.T) {
-	f := newForestDriver()
-
-	updateStringEntries(f, "first", "val")
-	topRootOf1, err1 := f.getTopRootHash()
-	updateStringEntries(f, "second", "val")
-	rootOfOneAfterSecondUpdate, err2 := f.getRootHash(1)
-
-	require.NoError(t, err1, "GetTopHash failed in first call")
-	require.NoError(t, err2, "GetTopHash failed in second call")
-	require.Equal(t, rootOfOneAfterSecondUpdate, topRootOf1, "getRootHash did not return expected hash")
-}
-
 func TestRootChangeAfterStateChange(t *testing.T) {
```

Comment:
second test can be adding to tree with root that is not exists ... and check that fails.

Reply:
added a test called TestUpdateTrieFailsForMissingBaseNode for this purpose.
the test required moving the allocation of the empty node hash again to inside the scope of NewForest() because inside the new test I was editing the value of the root hash. When using a global variable this caused this corruption to leak to other test.